### PR TITLE
Implementation of wyhash algorithm.

### PR DESCRIPTION
**Pull request by firestrand**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -68,8 +68,8 @@
                 <artifactId>maven-compiler-plugin</artifactId>
                 <version>3.8.0</version>
                 <configuration>
-                    <source>1.7</source>
-                    <target>1.7</target>
+                    <source>8</source>
```
Comment (leventov) [inline pom.xml]: Why need to change this?

File: src/main/java/net/openhft/hashing/LongHashFunction.java [hunk 2]
```diff
@@ -290,6 +290,33 @@ public static LongHashFunction xx(long seed) {
         return XxHash.asLongHashFunctionWithSeed(seed);
     }
 
+    /**
```
Comment (leventov) [inline src/main/java/net/openhft/hashing/LongHashFunction.java]: Please also add notes about the new algorithm to readme.md and package-info.java.

File: src/main/java/net/openhft/hashing/WyHash.java [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+
+    // Primes
+    public static final long _wyp0 = 0xa0761d6478bd642fL;
+    public static final long _wyp1 = 0xe7037ed1a0b428dbL;
+    public static final long _wyp2 = 0x8ebc6af09c88c6e3L;
+    public static final long _wyp3 = 0x589965cc75374cc3L;
+    public static final long _wyp4 = 0x1d8e4e27c47d124fL;
+
+    private long _wymum(long lhs, long rhs) {
+        //The Grade School method of multiplication is a hair faster in Java, primarily used here because the implementation is simpler.
```
Comment (leventov) [inline src/main/java/net/openhft/hashing/WyHash.java]: Please reformat code for lines shorter than 100 columns.

File: src/test/java/net/openhft/hashing/LongHashFunctionTest.java [hunk 4]
```diff
@@ -56,36 +58,48 @@ public static void testBoolean(LongHashFunction f, int len) {
...
 
     private static void testNegativePrimitives(LongHashFunction f) {
         byte[] bytes = new byte[8];
         Arrays.fill(bytes, (byte) -1);
-        assertEquals("byte hash", f.hashBytes(bytes, 0, 1), f.hashByte((byte) -1));
-        assertEquals("short hash", f.hashBytes(bytes, 0, 2), f.hashShort((short) -1));
-        assertEquals("char hash", f.hashBytes(bytes, 0, 2), f.hashChar((char) -1));
-        assertEquals("int hash", f.hashBytes(bytes, 0, 4), f.hashInt(-1));
-        assertEquals("long hash", f.hashBytes(bytes), f.hashLong(-1L));
+        long oneByteActual = f.hashBytes(bytes, 0, 1);
```
Comment (leventov) [inline src/test/java/net/openhft/hashing/LongHashFunctionTest.java]: Looks like "actual" params go first in this block, but last in the block above (lines 81-88)

File: pom.xml (hunk 1 above)
Comment (firestrand) [reply pom.xml]: IDE updated, I can revert.

Comment (leventov) [reply pom.xml]: Please revert

File: src/main/java/net/openhft/hashing/WyHash.java [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+    }
+    static LongHashFunction asLongHashFunctionWithSeed(long seed) {
+        return new AsLongHashFunctionSeeded(seed);
+    }
+
+    private static class AsLongHashFunctionSeeded extends AsLongHashFunction {
+        private static final long serialVersionUID = 0L;
+
+        private final long seed;
+        private final transient long voidHash = 0L;
```
Comment (leventov) [inline src/main/java/net/openhft/hashing/WyHash.java]: This field should be removed. BTW is it sure that wyhash's output for a zero-length sequence is 0 regardless of the seed?

File: src/main/java/net/openhft/hashing/WyHash.java [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+        public long seed() {
+            return 0L;
+        }
+
+        @Override
+        public long hashLong(long input) {
+            input = NATIVE_WY.toLittleEndian(input);
+            int hi = (int)(input & 0xFFFFFFFFL);
+            int lo = (int)((input >>> 32) & 0xFFFFFFFFL);
+            byte[] bytes = new byte[] {
```
Comment (leventov) [inline src/main/java/net/openhft/hashing/WyHash.java]: This is not zero-allocation. Please make a zero-allocation implementation.

File: src/main/java/net/openhft/hashing/WyHash.java [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+                (byte)Primitives.unsignedByte(lo >> 24)
+            };
+            return WyHash.INSTANCE.wyHash64(seed(), bytes,
+                                            UnsafeAccess.INSTANCE, BYTE_BASE, bytes.length);
+        }
+
+        @Override
+        public long hashInt(int input) {
+            input = NATIVE_WY.toLittleEndian(input);
+            byte[] bytes = new byte[] {
```
Comment (leventov) [inline src/main/java/net/openhft/hashing/WyHash.java]: And here

File: src/main/java/net/openhft/hashing/WyHash.java [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+                (byte)Primitives.unsignedByte(input >> 24)
+            };
+            return WyHash.INSTANCE.wyHash64(seed(), bytes,
+                                            UnsafeAccess.INSTANCE, BYTE_BASE, bytes.length);
+        }
+
+        @Override
+        public long hashShort(short input) {
+            input = NATIVE_WY.toLittleEndian(input);
+            byte[] hash = new byte[] {
```
Comment (leventov) [inline src/main/java/net/openhft/hashing/WyHash.java]: And here

File: src/main/java/net/openhft/hashing/WyHash.java [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+        }
+
+        @Override
+        public long hashChar(final char input) {
+            return hashShort((short)input);
+        }
+
+        @Override
+        public long hashByte(final byte input) {
+            byte[] bytes = new byte[]{input};
```
Comment (leventov) [inline src/main/java/net/openhft/hashing/WyHash.java]: And here

File: src/main/java/net/openhft/hashing/WyHash.java (hunk 5 above)
Comment (firestrand) [reply src/main/java/net/openhft/hashing/WyHash.java]: That is the behavior of the original C implementation, so matching the implementation output requires returning 0 for 0 input.
https://github.com/wangyi-fudan/wyhash/blob/master/wyhash.h#L35

**Comment by leventov** [comment]
@firestrand thank you. Released version 0.10 with your contribution included.

**Pull request merged**